### PR TITLE
Start using stack temporaries instead of allocated data

### PR DESCRIPTION
**Pull request by WalterBright**
Convert interpreter to using stack temporaries.

### CODE REVIEW

**Comment by WalterBright** [comment]
The purpose of this is to speed up CTFE and reduce its memory consumption. There's a lot more to come, this is just the first step.

File: src/dmd/dinterpret.d [hunk 1]
```diff
@@ -6230,11 +6249,26 @@ public:
     }
 }
 
-private Expression interpret(Expression e, InterState* istate, CtfeGoal goal = ctfeNeedRvalue)
+/********************************************
+ * Interpret the expression.
+ * Returns:
+ *    resulting expression
+ */
+
```
Comment (jacob-carlborg) [inline src/dmd/dinterpret.d]: No `Params` section.

File: src/dmd/dinterpret.d [hunk 2]
```diff
@@ -6230,11 +6249,26 @@ public:
...
+Expression interpret(Expression e, InterState* istate, CtfeGoal goal = ctfeNeedRvalue)
+{
+    UnionExp ue = void;
+    auto result = interpret(&ue, e, istate, goal);
+    if (result == ue.exp())
+        result = ue.copy();
+    return result;
+}
+
+Expression interpret(UnionExp* pue, Expression e, InterState* istate, CtfeGoal goal = ctfeNeedRvalue)
```
Comment (jacob-carlborg) [inline src/dmd/dinterpret.d]: No documentation.

File: src/dmd/dinterpret.d [hunk 3]
```diff
@@ -6248,11 +6282,20 @@ private Expression interpret(Expression e, InterState* istate, CtfeGoal goal = c
...
+Expression interpret(Statement s, InterState* istate)
+{
+    UnionExp ue = void;
+    auto result = interpret(&ue, s, istate);
+    if (result == ue.exp())
+        result = ue.copy();
+    return result;
+}
+
+Expression interpret(UnionExp* pue, Statement s, InterState* istate)
```
Comment (jacob-carlborg) [inline src/dmd/dinterpret.d]: No documentation.

**Review by jacob-carlborg — DISMISSED** [review]

**Review by RazvanN7 — APPROVED** [review]

**Comment by JinShil** [comment]
Just for my own education, how were any allocations saved with this PR?  I can only see that output was returned via a pointer parameter instead of returning it through `result`.

**Comment by WalterBright** [comment]
> how were any allocations saved with this PR?

Given `enum e = a + b + c;`, one allocation is saved.

**Pull request merged**